Web Inspector: Timelines: Unable to load more instances of a type in JavaScript Allocations view

rcaliman-apple opened the pull request:
#### 41ae46049e57b0879803ec62b8e814023f5b9e36
<pre>
Web Inspector: Timelines: Unable to load more instances of a type in JavaScript Allocations view
<a href="https://bugs.webkit.org/show_bug.cgi?id=278667">https://bugs.webkit.org/show_bug.cgi?id=278667</a>
<a href="https://rdar.apple.com/133810318">rdar://133810318</a>

Reviewed by BJ Burg.

For heap snapshot trees with many entries, instances are rendered batched in groups of 100.

A `WI.HeapSnapshotInstanceFetchMoreDataGridNode` with controls to generate the DOM for
the next batch or the full list is created, but it marked as `hidden` as a side-effect of
<a href="https://bugs.webkit.org/show_bug.cgi?id=157582.">https://bugs.webkit.org/show_bug.cgi?id=157582.</a>

A `WI.DataGrid` is virtualized to only display the limited number of data grid nodes
that are visible in the viewport. Any `WI.DataGridNode` marked as `hidden` is skipped
in `WI.DataGrid.updateVisibleRows()`:

```
let revealedRows = this._rows.filter((row) =&gt; row.revealed &amp;&amp; !row.hidden);
```

`WI.HeapSnapshotInstanceFetchMoreDataGridNode` gets marked as hidden when inserted inserted:

`WI.DataGrid.insertChild(node)` -&gt; `WI.DataGrid._applyFiltersToNodeAndDispatchEvent(node)`
-&gt; `WI.DataGrid._applyFiltersToNode(node)` where a delegate filtering method is called:
`WI.HeapSnapshotContentView.dataGridMatchNodeAgainstCustomFilters()`.

This delegate filter explicitly skips `WI.HeapSnapshotInstanceFetchMoreDataGridNode` nodes:

```
dataGridMatchNodeAgainstCustomFilters(node)
{
    console.assert(node);
    if (node instanceof WI.HeapSnapshotInstanceFetchMoreDataGridNode)
        return false;
    return true;
}
```

Combined with the virtualization logic from `WI.DataGrid.updateVisibleRows()` this results
in always hiding the controls to load more entries.

The intent was likely to skip showing these controls when there are no matches for a filter query.

This patch ensures that the delegate filter checks for visible sibling data grid nodes when deciding
whether to show the controls to load more entries.

The filter query matches against the top-level instance type, so the number of matched individual instances
of that type will not vary. Therefore, the controls to load more entries don&apos;t have to update their counts.

* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotContentView.js:
(WI.HeapSnapshotContentView.prototype.dataGridMatchNodeAgainstCustomFilters.hasVisibleSiblingNodes):
(WI.HeapSnapshotContentView.prototype.dataGridMatchNodeAgainstCustomFilters):

Canonical link: <a href="https://commits.webkit.org/283215@main">https://commits.webkit.org/283215@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c94eaf4ed63c14eed2a73a11802718356d207b1e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65229 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44598 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69253 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15835 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16117 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52404 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10963 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68295 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41223 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56463 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33027 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37894 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13834 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-masking/clip-path/animations/clip-path-animation-font-size-mixed-change.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14712 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59781 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14174 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70958 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9181 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13651 "Found 1 new test failure: imported/w3c/web-platform-tests/media-source/mediasource-activesourcebuffers.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59729 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9213 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56524 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60004 "Found 2 new API test failures: /WebKitGTK/TestResources:/webkit/WebKitWebResource/get-data-error, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/basic-hierarchy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14458 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7588 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1263 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40408 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41485 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42666 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41229 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ba5dded07da820ce30fb7a116cbc1011c52c4734. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/32721

the-chenergy commented on Source/WebInspectorUI/UserInterface/Views/HeapSnapshotContentView.js:
```diff
@@ -72,7 +72,16 @@ WI.HeapSnapshotContentView = class HeapSnapshotContentView extends WI.ContentVie
+            return false;
+        }
+
+        if (node instanceof WI.HeapSnapshotInstanceFetchMoreDataGridNode && !hasVisibleSiblingNodes())
```
Maybe the `siblingNode` for-loop is simple enough that we don't need the lambda?
```js
    dataGridMatchNodeAgainstCustomFilters(node)
    {
        console.assert(node);

        if (!(node instanceof WI.HeapSnapshotInstanceFetchMoreDataGridNode))
            return true;

        for (let siblingNode of node.parent.children) {
            if (!siblingNode.hidden && siblingNode !== node)
                return true;
        }

        return false;
    }
```

rcaliman-apple replied:
I think I prefer the explicitly named function call.

The `WI.HeapSnapshotInstanceFetchMoreDataGridNode` entry should be hidden *only* when there are no visible siblings, regardless of conditions.

Consider these cases:
- no filtering, many nodes
- filtering, no matching nodes
- filtering, matching many nodes

Reading `node instanceof WI.HeapSnapshotInstanceFetchMoreDataGridNode && !hasVisibleSiblingNodes()` is clearer about that intent.

dcrousso replied:
IMO the inlined version is nicer to read as it keeps the logic more localized (i.e. we only really care about this code in the case that we're dealing with a `WI.HeapSnapshotInstanceFetchMoreDataGridNode` so "polluting" the outside scope is somewhat not ideal)

but putting that aside for a moment

i don't think this is necessarily the correct fix

in other places that do things like `dataGridMatchNodeAgainstCustomFilters`, usually the first check is to even see if there is an active filter, which i think in this case can be checked via `this._dataGrid.filterText`

as far as still showing it when filtering, i think we may just want to successively iterate over `.previousSibling` instead of grabbing the `.parent.children` since we know that it will always be the last child (and ideally we'd make this a member function of `WI.HeapSnapshotInstanceFetchMoreDataGridNode` since that's really where this logic should live)

rcaliman-apple replied:
> in other places that do things like dataGridMatchNodeAgainstCustomFilters, usually the first check is to even see if there is an active filter, which i think in this case can be checked via this._dataGrid.filterText

The existence or absence of a filter text is not relevant to showing a `WI.HeapSnapshotInstanceFetchMoreDataGridNode` entry.

Consider the case where even with a filter there are still more results than the batch limit. (Filtering does not operate on the entire list of results, it operates only on created nodes).

The only thing that is truly relevant is whether there are more than `WI.HeapSnapshotClassDataGridNode.ChildrenBatchLimit` sibling nodes visible. But we don't have to be check that precisely.

The filter text matches only against the entries for the column names of the `WI.DataGrid`:
(`className` is the relevant column name for `WI.HeapSnapshotInstanceDataGridNode` and `WI.HeapSnapshotClassDataGridNode`)

https://github.com/WebKit/WebKit/blob/2cf8384747d76adddf7eba0ff02683214534570f/Source/WebInspectorUI/UserInterface/Views/DataGrid.js#L492

https://github.com/WebKit/WebKit/blob/8f07bf9963e1bf4e307fc0d68556df4b21d2e95d/Source/WebInspectorUI/UserInterface/Views/DataGridNode.js#L169-L194

An instance type (`className`) that has more results than the limit will *still* have more results than the limit even when matched by the filter text.

The filter text isn't currently matched against more granular data within each instance. 
This means we can return early for _any_ visible sibling nodes instead of having to count them.

https://github.com/user-attachments/assets/f5ace730-294c-45b3-baba-660368e38fc8

> as far as still showing it when filtering, i think we may just want to successively iterate over .previousSibling instead of grabbing the .parent.children since we know that it will always be the last child (and ideally we'd make this a member function of WI.HeapSnapshotInstanceFetchMoreDataGridNode since that's really where this logic should live)

This is a good idea. 👍  

I updated the patch to have `WI.HeapSnapshotInstanceFetchMoreDataGridNode.hasVisibleSiblingNodes()`. It satisfies both the desires to see the check inline ...

```
if (node instanceof WI.HeapSnapshotInstanceFetchMoreDataGridNode && !hasVisibleSiblingNodes())
```

... and the desire to explicitly name the method for clarity.

dcrousso replied:
> The existence or absence of a filter text is not relevant to showing a `WI.HeapSnapshotInstanceFetchMoreDataGridNode` entry.

it absolutely is.  if there's no filter text then there's no reason to filter out `WI.HeapSnapshotInstanceFetchMoreDataGridNode` regardless of its or sibling state.  the entire point of `dataGridMatchNodeAgainstCustomFilters` is to support filtering beyond just text (e.g. a button that has some special functionality).  having something like this at the beginning
```js
if (!this._dataGrid.filterText)
    return true;
```
will be a much quicker short-circuit than having to do an `instanceof` check and look at siblings

rcaliman-apple replied:
> the entire point of dataGridMatchNodeAgainstCustomFilters is to support filtering beyond just text (e.g. a button that has some special functionality). having something like this at the beginning

If that was truly the point and  `dataGridMatchNodeAgainstCustomFilters` was an additive filter, it should not be called if there is no filter text. But that's not what I'm getting from your suggestion or the implementation:

https://github.com/WebKit/WebKit/blob/ef7ef86e4cf56ad91d0a06677ba8de5de8cc3f1e/Source/WebInspectorUI/UserInterface/Views/DataGrid.js#L487-L498

`WI.DataGrid._applyFiltersToNode(node)` is intentionally designed to `matchTextFilter()` if there is no filter text which is then used in conjunction with https://github.com/WebKit/WebKit/blob/ef7ef86e4cf56ad91d0a06677ba8de5de8cc3f1e/Source/WebInspectorUI/UserInterface/Views/DataGrid.js#L520

Seems odd to have to recheck filter text in the delegate function if the point of the `dataGridMatchNodeAgainstCustomFilters` is to _further_ filter results.

dcrousso replied:
it's not additive, it's alternative (hence the "custom" name)

the idea is that it can be used to do other kinds of filtering beyond just matching text

for example, the Sources Tab has a button next to the text filter that will cause only resources with issues to be shown.  it doesn't matter whether the developer has entered anything into the text filter in order for that button to have an effect.  so if the Sources Tab used a `WI.DataGrid` then it would need functionality like `dataGridMatchNodeAgainstCustomFilters` in order to have that effect

another example is the resource type `WI.ScopeBar` in the Network Tab.  back when that used a `WI.DataGrid` (instead of the `WI.Table` it currently uses) it likely would've used `dataGridMatchNodeAgainstCustomFilters` to implement that kind of custom filtering (again regardless of whether the developer had input anything into the URL filter)

in fact, `dataGridMatchNodeAgainstCustomFilters` being called regardless of whether there's a text filter is a critical part of how `WI.TimelineView` works as without that we'd need _yet another_ mechanism for ensuring that only timeline entries within the selected time range are shown

in this case, you're using `dataGridMatchNodeAgainstCustomFilters` to do an additional filter step beyond just the text filter (i.e. there's no button or `WI.ScopeBar` that has filtering functionality completely independent of the text filter), so we should check that the developer is doing a text filter before bothering to do anything else

dcrousso replied:
also ive been thinking about this a bit more, and i think there needs to be a FIXME here

ideally, we'd only ever show `WI.HeapSnapshotInstanceFetchMoreDataGridNode` if we know for a fact that something hidden matches the (text) filter, as otherwise we could be misleading the developer into thinking that when it's not true

similarly, if we're only showing `WI.HeapSnapshotInstanceFetchMoreDataGridNode` if one of its previous siblings matches the (text) filter, then in situations where none do we would not show the `WI.HeapSnapshotInstanceFetchMoreDataGridNode` even though something hidden by it might match the (text) filter

as such, what we really need is a way to search the entire heap snapshot for matches so that we can gather all of the _data_ to show instead of using `WI.DataGrid` logic to just search what's in the UI

i think this should probably be a followup though, as that's likely going to be much more complicated, and in the meantime we should fix the bug of allowing `WI.HeapSnapshotInstanceFetchMoreDataGridNode` to be shown when not filtering for anything

rcaliman-apple replied:
I disagree with you, but I added the check for the filter text anyway since it won't affect the behavior of this patch at all.

The examples you provided are indeed about filtering results. They're orthogonal to the issue here, where filtering is used as a hack IMO to toggle the visibility of a UI control `WI.HeapSnapshotInstanceFetchMoreDataGridNode` that happens be mixed in with the data nodes.

_If_ filtering would work against the full dataset *and* granularly, and I agree that's a better experience, that would mean toggling the visibility of the `WI.HeapSnapshotInstanceFetchMoreDataGridNode` based on content. Whether or not a filter text is present would still be irrelevant; the number of filtered results matters. 

> it's not additive, it's alternative (hence the "custom" name)
> the idea is that it can be used to do other kinds of filtering beyond just matching text

If that's true, then this becomes that: 
```diff
- if (matchTextFilter() && this.matchNodeAgainstCustomFilters(node))
+ if (matchTextFilter() || this.matchNodeAgainstCustomFilters(node))
```

But since we're pretending that results with no filterable data did match, we do this check in the delegate:

```
if (!this._dataGrid.filterText)
  return true;
```

Either way, this patch serves its purpose with and without this addition so I'm putting it in.

dcrousso replied:
> The examples you provided are indeed about filtering results. They're orthogonal to the issue here, where filtering is used as a hack IMO to toggle the visibility of a UI control `WI.HeapSnapshotInstanceFetchMoreDataGridNode` that happens be mixed in with the data nodes.

this bug was caused by [248198@main](https://commits.webkit.org/214193@main) which specifically added support for filtering, so im not really sure how this is not related to filtering

also btw `WI.HeapSnapshotInstanceFetchMoreDataGridNode` is only added if there are more than 100 values, so it only exists in those cases (i.e. it's not like it's always added and selectively hidden based on whether or not there are more things to show)

the ideal behavior is that if no filter text is provided, these would not be filtered out at all, so the only situation in which they should be hidden is if the filter text doesn't match

> _If_ filtering would work against the full dataset _and_ granularly, and I agree that's a better experience, that would mean toggling the visibility of the `WI.HeapSnapshotInstanceFetchMoreDataGridNode` based on content. Whether or not a filter text is present would still be irrelevant; the number of filtered results matters.

the filter text would be what determines wither the `WI.HeapSnapshotInstanceFetchMoreDataGridNode` is shown.  if there's no filter text then it should always be shown.  if there is filter text then it would only be shown if something hidden by it matches (or in the case of this PR, something else before it matches).

> > it's not additive, it's alternative (hence the "custom" name)
> > the idea is that it can be used to do other kinds of filtering beyond just matching text
> 
> If that's true, then this becomes that:
> 
> ```diff
> - if (matchTextFilter() && this.matchNodeAgainstCustomFilters(node))
> + if (matchTextFilter() || this.matchNodeAgainstCustomFilters(node))
> ```
> 
> But since we're pretending that results with no filterable data did match, we do this check in the delegate:
> 
> ```
> if (!this._dataGrid.filterText)
>   return true;
> ```

i agree that the existing code isn't written super well, but it's effectively doing what you describe because `matchTextFilter()` will always be `true` if there is no filter text (because `this._textFilterRegex` will not be set)

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0e282c0cabcd2a1e65fc8b4547a865f36f9f25bb. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/32721

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c94eaf4ed63c14eed2a73a11802718356d207b1e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/32721

rcaliman-apple committed:
Web Inspector: Timelines: Unable to load more instances of a type in JavaScript Allocations view
https://bugs.webkit.org/show_bug.cgi?id=278667
rdar://133810318

Reviewed by BJ Burg.

For heap snapshot trees with many entries, instances are rendered batched in groups of 100.

A `WI.HeapSnapshotInstanceFetchMoreDataGridNode` with controls to generate the DOM for
the next batch or the full list is created, but it marked as `hidden` as a side-effect of
https://bugs.webkit.org/show_bug.cgi?id=157582.

A `WI.DataGrid` is virtualized to only display the limited number of data grid nodes
that are visible in the viewport. Any `WI.DataGridNode` marked as `hidden` is skipped
in `WI.DataGrid.updateVisibleRows()`:

```
let revealedRows = this._rows.filter((row) => row.revealed && !row.hidden);
```

`WI.HeapSnapshotInstanceFetchMoreDataGridNode` gets marked as hidden when inserted inserted:

`WI.DataGrid.insertChild(node)` -> `WI.DataGrid._applyFiltersToNodeAndDispatchEvent(node)`
-> `WI.DataGrid._applyFiltersToNode(node)` where a delegate filtering method is called:
`WI.HeapSnapshotContentView.dataGridMatchNodeAgainstCustomFilters()`.

This delegate filter explicitly skips `WI.HeapSnapshotInstanceFetchMoreDataGridNode` nodes:

```
dataGridMatchNodeAgainstCustomFilters(node)
{
    console.assert(node);
    if (node instanceof WI.HeapSnapshotInstanceFetchMoreDataGridNode)
        return false;
    return true;
}
```

Combined with the virtualization logic from `WI.DataGrid.updateVisibleRows()` this results
in always hiding the controls to load more entries.

The intent was likely to skip showing these controls when there are no matches for a filter query.

This patch ensures that the delegate filter checks for visible sibling data grid nodes when deciding
whether to show the controls to load more entries.

The filter query matches against the top-level instance type, so the number of matched individual instances
of that type will not vary. Therefore, the controls to load more entries don't have to update their counts.

* Source/WebInspectorUI/UserInterface/Views/HeapSnapshotContentView.js:
(WI.HeapSnapshotContentView.prototype.dataGridMatchNodeAgainstCustomFilters.hasVisibleSiblingNodes):
(WI.HeapSnapshotContentView.prototype.dataGridMatchNodeAgainstCustomFilters):

Canonical link: https://commits.webkit.org/283215@main

webkit-commit-queue commented:
Committed 283215@main (41ae46049e57): <https://commits.webkit.org/283215@main>

Reviewed commits have been landed. Closing PR #32721 and removing active labels.
